zerver/lib/actions.py

```diff
@@ -2249,55 +2152,55 @@ def send_pm_if_empty_stream(stream: Optional[Stream],
 
     send_rate_limited_pm_notification_to_bot_owner(sender, realm, content)
 
-def validate_sender_can_write_to_stream(sender: UserProfile,
-                                        stream: Stream,
-                                        forwarder_user_profile: Optional[UserProfile]) -> None:
-    # Our caller is responsible for making sure that `stream` actually
-    # matches the realm of the sender.
-
-    # Organization admins can send to any stream, irrespective of the stream_post_policy value.
-    if sender.is_realm_admin or is_cross_realm_bot_email(sender.delivery_email):
-        pass
-    elif sender.is_bot and (sender.bot_owner is not None and
-                            sender.bot_owner.is_realm_admin):
-        pass
-    elif stream.stream_post_policy == Stream.STREAM_POST_POLICY_ADMINS:
-        raise JsonableError(_("Only organization administrators can send to this stream."))
-    elif stream.stream_post_policy == Stream.STREAM_POST_POLICY_RESTRICT_NEW_MEMBERS:
-        if sender.is_bot and (sender.bot_owner is not None and
-                              sender.bot_owner.is_new_member):
-            raise JsonableError(_("New members cannot send to this stream."))
-        elif sender.is_new_member:
-            raise JsonableError(_("New members cannot send to this stream."))
-
-    if not (stream.invite_only or sender.is_guest):
-        # This is a public stream and sender is not a guest user
-        return
-
-    if subscribed_to_stream(sender, stream.id):
-        # It is private, but your are subscribed
-        return
-
-    if sender.is_api_super_user:
-        return
-
-    if (forwarder_user_profile is not None and forwarder_user_profile.is_api_super_user):
-        return
-
-    if sender.is_bot and (sender.bot_owner is not None and
-                          subscribed_to_stream(sender.bot_owner, stream.id)):
-        # Bots can send to any stream their owner can.
-        return
-
-    if sender.delivery_email == settings.WELCOME_BOT:
-        # The welcome bot welcomes folks to the stream.
-        return
-
-    if sender.delivery_email == settings.NOTIFICATION_BOT:
-        return
-
-    # All other cases are an error.
-    raise JsonableError(_("Not authorized to send to stream '%s'") % (stream.name,))
+# def validate_sender_can_write_to_stream(sender: UserProfile,
```

Comment:
There's no need to leave the old code commented out here.  Just move it.

A better thing to to here is to break down the code moves into smaller commits.  For example, you can remove just the "leaf" functions in the call tree first as small, coherent commits.